### PR TITLE
adding interface for custom responses

### LINKED ISSUE #10960: Implement custom response interface for service workers
**Note**: This is part of the [Service Workers](https://github.com/servo/servo/wiki/Summer-of-Code-2016:-ServiceWorker-infrastructure) project, and assigned to myself.

For scopes which are controlled by a service worker to be able to respond with a custom response, the `http_loader.rs` needs a bit tweaking, to be able to know the source of the request. The network code needs to ask for custom response from client, then wrap it in a `StreamedResponse`.

I am currently working on it.

### PR DESCRIPTION
**Pull request by creativcoder**
Fixes #10960 

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10961)

### CODE REVIEW

File: components/net_traits/lib.rs [hunk 1]
```diff
@@ -91,6 +155,8 @@ pub struct LoadData {
     /// The policy and referring URL for the originator of this request
     pub referrer_policy: Option<ReferrerPolicy>,
     pub referrer_url: Option<Url>,
+    #[ignore_heap_size_of = "Defined in hyper"]
```
Comment (nox) [inline components/net_traits/lib.rs]: Hyper has `HeapSizeOf` support, and that `RequestSource` enum isn't defined in Hyper.

**Comment by creativcoder** [comment]
@jdm r?

File: components/net_traits/lib.rs (hunk 1 above)
Comment (creativcoder) [inline components/net_traits/lib.rs]: Yeah, this was an issue i had, when i put derive HeapSizeOf attribute on RequestSource, due to the IpcSender wrapped in it. Do i have to make an impl HeapSizeOf for IpcSender ?

**Comment by jdm** [comment]
We should modify `PendingAsyncLoad::new` to accept a `RequestSource` argument, as well as modify `XMLHttpRequest.rs`'s use of LoadData::new, and `Window::load_url`... It's probably time to make `LoadData::new` accept a trait object argument that can be used to obtain the request source, as well as referrer and referrer policy, and pipeline.

-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[components/net/http_loader.rs, line 483 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW6o6wUfJPTfEUgc6N:-KHW6o6x9GG89bXUb49G:-1310392102) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net/http_loader.rs#L483)):*
> ```Rust
>     }
> 
>     fn custom_response(m: Metadata, response: Box<HttpResponse>) -> StreamedResponse {
> ```

We should be able to use `from_http_response` as far as I can tell, and that will properly support custom responses that need decoding.

---

*[components/net/http_loader.rs, line 760 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW7A27ev6O0U_2wdkR:-KHW7A28Jq49DxBpJh_w:1142249545) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net/http_loader.rs#L760)):*
> ```Rust
>     let (msg_sender, msg_receiver) = ipc::channel().unwrap();
>     match load_data.source {
>         RequestSource::Window(ref sender) => {
> ```

It looks like we can merge these two cases together.

---

*[components/net_traits/lib.rs, line 130 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW7uQxiKKnM1_KZlkV:-KHW7uQxiKKnM1_KZlkW:-707643990) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net_traits/lib.rs#L130)):*
> ```Rust
> }
> 
> pub struct ReadableCustomResponse {
> ```

Let's move this into `http_loader.rs`, which will let us keep `HttpResponse` there as well. `to_readable_response` can be a standalone function there as well if we make the fields of CustomResponse public.

---

*[components/net_traits/lib.rs, line 155 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW8DtqqLuSimQpqnsP:-KHW8DtriFN4GWfwajE5:-1502145506) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net_traits/lib.rs#L155)):*
> ```Rust
> #[derive(Clone, Deserialize, Serialize, HeapSizeOf)]
> pub enum RequestSource {
>     Window(#[ignore_heap_size_of = "Defined in Std"] CustomResponseSender),
> ```

These are define in ipc-channel instead.

---

*[components/net_traits/lib.rs, line 206 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW8RbxYmQd8w3v4YUa:-KHW8RbxYmQd8w3v4YUb:1366238341) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net_traits/lib.rs#L206)):*
> ```Rust
>     }
> 
>     pub fn set_load_source(&mut self, req_src: RequestSource) {
> ```

This method probably isn't necessary since the field is public.

---

*[components/script/script_thread.rs, line 211 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW97JsLBkVz3r3IPgp:-KHW97JsLBkVz3r3IPgq:-1197430468) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/script/script_thread.rs#L211)):*
> ```Rust
>     FromImageCache(ImageCacheResult),
>     FromScheduler(TimerEvent),
>     FromServiceWorker(IpcSender<Option<CustomResponse>>),
> ```

This isn't a message from a service worker; it's from a network request.

---

*[components/script/script_thread.rs, line 330 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW8pBjDDdwMvt0tuDI:-KHW8pBjDDdwMvt0tuDJ:338906494) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/script/script_thread.rs#L330)):*
> ```Rust
>     custom_message_chan: CustomResponseSender,
> 
>     custom_message_port: Receiver<IpcSender<Option<CustomResponse>>>,
> ```

Documentation for these would be nice.

---

*[components/script/script_thread.rs, line 1021 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW90FbmUKLdELIhb_s:-KHW90FbmUKLdELIhb_t:1839427290) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/script/script_thread.rs#L1021)):*
> ```Rust
>         // we may detect controlling service workers here
>         // We send None as default
>         msg.send(None).unwrap();
> ```

Let's use `let _ msg.send(None);` instead.

---

*[components/script/script_thread.rs, line 1926 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW9H9W049gMC3tDszG:-KHW9H9XNrKK98XX9cVn:-2072554518) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/script/script_thread.rs#L1926)):*
> ```Rust
>         }
> ```

nit: remove this extra newline.

---

*[components/script/dom/window.rs, line 309 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW9Pu3i6x4kdaFi-SP:-KHW9Pu4S2XNFKwNTm8z:-2072554518) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/script/dom/window.rs#L309)):*
> ```Rust
>         self.custom_message_chan.clone()
>     }
> ```

nit: remove this extra newline.

---

*[tests/unit/net/http_loader.rs, line 1674 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW9hqHZlGjBnatLpHW:-KHW9hqHZlGjBnatLpHX:505322894) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1674)):*
> ```Rust
>         Headers::new(),
>         RawStatus(200, Cow::Borrowed("Ok")),
>         Vec::<u8>::new()
> ```

This should contain a real body that we verify is received.

---

*[tests/unit/net/http_loader.rs, line 1686 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHW9woI0rkgu49WWBLv:-KHW9woI0rkgu49WWBLw:-1554806584) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1686)):*
> ```Rust
>     load_data.method = Method::Get;
>     let mut content_length = Headers::new();
>     content_length.set(ContentLength(0));
> ```

I don't think any of the modifications here are necessary.

---

*[tests/unit/net/http_loader.rs, line 1694 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWAGRI-xyT9P6sYrKW:-KHWAGRJPB4IvEakIooj:1552350583) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1694)):*
> ```Rust
>         let _ = load(
>         &load_data.clone(), &ui_provider, &http_state,
>         None, &AssertMustIncludeHeadersRequestFactory {
> ```

Let's create a factory that returns a MockResponse with a 404 status instead of using `AssertMustIncludeHeadersRequestFactory`.

---

*[tests/unit/net/http_loader.rs, line 1698 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWA4M4L9NoZ1u_s-Uj:-KHWA4M4L9NoZ1u_s-Uk:732401897) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1698)):*
> ```Rust
>             body: <[_]>::to_vec(&[])
>         }, DEFAULT_USER_AGENT.to_owned(), &CancellationListener::new(None));
>     });
> ```

Let's send the result of calling load on a channel, and receive it below after sending the mock response. This will allow us to assert that the response body matches the mock response we provided.

---

*[tests/unit/net/http_loader.rs, line 1705 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWAZiGUW9N0kHtZZKY:-KHWAZiJbiYZjpBwfep1:2103053444) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1705)):*
> ```Rust
> 
> #[test]
> fn test_custom_response_from_window() {
> ```

We should be able to move the previous test into a function that accepts a `RequestSource` argument and uses it. We can then call that function for each source that we want to test without duplicating all the code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KHWB3lyGBYk7Os7fKez:-273921067)*

**Comment by creativcoder** [comment]
Will be adding the trait object arg, in subsequent commit.

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KH_4iS1GCDFClK7qzuQ:-2052561454)*

**Comment by creativcoder** [comment]
*[tests/unit/net/http_loader.rs, line 1698 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWA4M4L9NoZ1u_s-Uj:-KH_54GidDLq4XLk_-Hc:1614936314) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1698)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> Let's send the result of calling load on a channel, and receive it below after sending the mock response. This will allow us to assert that the response body matches the mock response we provided.</details>
This needed an `unsafe impl Send for StreamedResponse {}` in lib.rs. But this unsafe impl is only needed for the tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961)*

**Comment by creativcoder** [comment]
*[tests/unit/net/http_loader.rs, line 1705 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWAZiGUW9N0kHtZZKY:-KH_5cX0bJCZIrGTwawG:-1055111851) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1705)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> We should be able to move the previous test into a function that accepts a `RequestSource` argument and uses it. We can then call that function for each source that we want to test without duplicating all the code.</details>
I found adding an enum, to be simple, instead of returning JoinHandle from `load_request_with_source`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961)*

File: components/net/lib.rs [hunk 2]
```diff
@@ -10,7 +10,7 @@
 #![feature(plugin)]
 #![plugin(plugins)]
 
-#![deny(unsafe_code)]
+//#![deny(unsafe_code)]
```
Comment (creativcoder) [inline components/net/lib.rs]: Not sure, whether we use unsafe for StreamedResponse, which is required only for unit test.

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, creativcoder (Rahul Sharma) wrote…</i></summary>
> Will be adding the trait object arg, in subsequent commit.</details>


---

Reviewed 9 of 12 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net/http_loader.rs, line 178 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_HEcJPBbyACFRNcUS:-KH_HEcJPBbyACFRNcUT:-1719556325) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/components/net/http_loader.rs#L178)):*
> ```Rust
> }
> 
> unsafe impl Send for StreamedResponse {}
> ```

We should be able to remove this with the other changes I propose.

---

*[components/net/lib.rs, line 13 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_4pLg8OUSXqq44NoL-r3-13:-KH_HVxRFC56C94TLose:-183018310) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/components/net/lib.rs#L13)):*
<details><summary><i>Previously, creativcoder (Rahul Sharma) wrote…</i></summary>
> Not sure, whether we use unsafe for StreamedResponse, which is required only for unit test.</details>
Any use of unsafe can have `#[allow(unsafe_code)]` instead of disabling this globally.

---

*[components/net_traits/lib.rs, line 14 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_HhFUgzAGf5Ze3k3P:-KH_HhFUgzAGf5Ze3k3Q:753395594) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/components/net_traits/lib.rs#L14)):*
> ```Rust
> #![plugin(heapsize_plugin, serde_macros)]
> 
> #![deny(unsafe_code)]
> ```

Why this change?

---

*[components/net_traits/lib.rs, line 317 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_I8ZY7nZPpdYcfhU7:-KH_I8ZY7nZPpdYcfhU8:-1885511881) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/components/net_traits/lib.rs#L317)):*
> ```Rust
>             referrer_policy: referrer_policy,
>             referrer_url: referrer_url,
>             source: source
> ```

This doesn't seem to be used yet?

---

*[components/net_traits/lib.rs, line 325 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_Htu_CV-2NQ-QEko3:-KH_Htu_CV-2NQ-QEko4:2142847647) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/components/net_traits/lib.rs#L325)):*
> ```Rust
>         self.guard.neuter();
>         let load_data = LoadData::new(self.context,
>                                         self.url,
> ```

nit: align these under the `self` on the first line, please.

---

*[components/script/document_loader.rs, line 143 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_IDFmG3XsR6ET7zuw:-KH_IDFmG3XsR6ET7zux:1054008724) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/components/script/document_loader.rs#L143)):*
> ```Rust
>                               referrer.get_referrer_policy(),
>                               Some(referrer.url().clone()),
>                               RequestSource::None)
> ```

This should be a `RequestSource::Window`.

---

*[tests/unit/net/http_loader.rs, line 1698 \[r2\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHWA4M4L9NoZ1u_s-Uj:-KH_IuY05J6F8XtYLTzG:594514921) ([raw file](https://github.com/servo/servo/blob/145e8fd5ad5938853de445749f49d8795dffbf43/tests/unit/net/http_loader.rs#L1698)):*
<details><summary><i>Previously, creativcoder (Rahul Sharma) wrote…</i></summary>
> This needed an `unsafe impl Send for StreamedResponse {}` in lib.rs. But this unsafe impl is only needed for the tests.</details>
We can do
```rust
let mut response = load(...);
let metadata = response.metadata.clone();
let body = read_response(&mut response);
(metadata, body)
```
and avoid the need for the unsafe impl.

---

*[tests/unit/net/http_loader.rs, line 1726 \[r3\]](https://reviewable.io:443/reviews/servo/servo/10961#-KH_JntuJPNFexSURLrq:-KH_Jntvu_IJ6RBgk5yy:-1198629754) ([raw file](https://github.com/servo/servo/blob/775605d3c9a5ce589108a171c7c451167c532161/tests/unit/net/http_loader.rs#L1726)):*
> ```Rust
> #[test]
> fn test_custom_response_from_window() {
>     let expected_body = <[_]>::to_vec("Yay! From Window".as_bytes());
> ```

We'll want to check the 200 status as well. This may also be possible to write as `b"Yay! From Window"`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KH_K4K7c0BGT4apfcE-:1458266033)*

File: components/net/lib.rs (hunk 2 above)
Comment (Ms2ger) [inline components/net/lib.rs]: This is not acceptable, and AFAICT not even necessary.

**Comment by jdm** [comment]
Looking good!
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---

Reviewed 15 of 20 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/net_traits/lib.rs, line 148 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtk1Fsks--QnoZotZu:-KHtk1FtnRfrF-K0B2l6:794671085) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/components/net_traits/lib.rs#L148)):*
> ```Rust
> 
> pub trait LoadOrigin {
>     fn referrer_url(&self) -> Option<Url> {
> ```

I would prefer not to have default implementations of these methods; we usually do not want None or default values here.

---

*[components/script/document_loader.rs, line 148 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtkJ8wLswi0cUdtpSV:-KHtkJ8wLswi0cUdtpSW:1127047147) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/components/script/document_loader.rs#L148)):*
> ```Rust
>                               referrer.get_referrer_policy(),
>                               Some(referrer.url().clone()),
>                               RequestSource::Window(client_chan))
> ```

We should be able to obtain the client_chan via the Document parameter instead of passing it in directly.

---

*[components/script/dom/dedicatedworkerglobalscope.rs, line 255 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtl3wYsn36c8TlaaqP:-KHtkontcQzlhHSYp6Lb:523320481) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/components/script/dom/dedicatedworkerglobalscope.rs#L255)):*
> ```Rust
>             *main_thread_rt.lock().unwrap() = Some(SharedRt::new(&runtime));
> 
>             let (msg_chan, msg_port) = ipc::channel().unwrap();
> ```

This could probably be created directly in WorkerGlobalScope::new_inherited.

---

*[components/script/dom/workerglobalscope.rs, line 221 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtlOX6vqSmHY-YRH6L:-KHtlOX6vqSmHY-YRH6M:1188481747) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/components/script/dom/workerglobalscope.rs#L221)):*
> ```Rust
>     }
>     fn pipeline_id(&self) -> Option<PipelineId> {
>         None
> ```

We should return self.pipeline() here.

---

*[tests/unit/net/http_loader.rs, line 1536 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtm8AnzdY2M9ieAV7S:-KHtm8AocL0K7yus3S2l:-214981563) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/tests/unit/net/http_loader.rs#L1536)):*
> ```Rust
> }
> 
> fn assert_referer_header_matches(origin_info: &LoadOrigin, request_url: &str, expected_referrer: &str) {
> ```

Let's keep these on separate lines.

---

*[tests/unit/net/http_loader.rs, line 1557 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtmcF7x46xdtMSoDcV:-KHtmcF7x46xdtMSoDcW:-1337818002) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/tests/unit/net/http_loader.rs#L1557)):*
> ```Rust
> }
> 
> struct NoReferrer<'a> {
> ```

How is this different than LoadOriginInfo?

---

*[tests/unit/net/http_loader.rs, line 1573 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtmKOObVB6HzERMTvJ:-KHtmKOPEES-wPyAmt-f:1247654785) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/tests/unit/net/http_loader.rs#L1573)):*
> ```Rust
>     }
> fn pipeline_id(&self) -> Option<PipelineId> {
>     None
> ```

nit: indentation here is wonky.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KHtms31NZlG3kAHmkTJ:-1485515698)*

**Comment by creativcoder** [comment]
*[tests/unit/net/http_loader.rs, line 1557 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHtmcF7x46xdtMSoDcV:-KHu7fhfaMMC9mXnh9Qd:-372457293) ([raw file](https://github.com/servo/servo/blob/709c4f2f12d722438e2e9568420fd42ce2785b16/tests/unit/net/http_loader.rs#L1557)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> How is this different than LoadOriginInfo?</details>
Whoops ! on previous iterations of the code; they had different fields; anyways removed that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961)*

**Comment by jdm** [comment]
Very close!
-S-awaiting-review +S-needs-code-changes 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---

Reviewed 10 of 10 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/dedicatedworkerglobalscope.rs, line 157 \[r6\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHuctgnBg3cYMjA2EKu:-KHucXXNYdXX2HVkRW08:480849189) ([raw file](https://github.com/servo/servo/blob/eb950b791bdf636ff14b0801e5020ae714cf0583/components/script/dom/dedicatedworkerglobalscope.rs#L157)):*
> ```Rust
> }
> 
> impl LoadOrigin for TrustedWorkerAddress {
> ```

Rather than implementing this for TrustedWorkerAddress, we should have a new struct called WorkerScriptLoadOrigin that is passed in to `run_worker_scope` as an argument and returns the values it contains.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KHudd93RJ7H1Tp8ayrU:-140489254)*

**Comment by jdm** [comment]
@bors-servo: r+

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---

Reviewed 1 of 2 files at r7, 5 of 5 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KHvMsNYV_XKSxvc8Gbj:-317921516)*

**Comment by jdm** [comment]
Oh, I'm pretty sure we have a problem where the custom message receiver is never being read from in workers, so any network requests initiated from workers never complete.

**Comment by jdm** [comment]
-S-awaiting-review +S-needs-code-changes +S-fails-tidy 

<details><summary><i>Previously, highfive wrote…</i></summary>
> New code was committed to pull request.</details>


---

Reviewed 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[components/net/data_loader.rs, line 84 \[r9\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHz5P-y2sVxDXa2d7lL:-KHz5P-zM8KQnFGNzrIH:2015259436) ([raw file](https://github.com/servo/servo/blob/dd4baae890fa6a22c1e33f779571f02f6c77af9e/components/net/data_loader.rs#L84)):*
> ```Rust
>     }
> 
>     let (msg_sender, msg_receiver) = ipc::channel().unwrap();
> ```

Custom responses are only handled for HTTP requests: https://fetch.spec.whatwg.org/#http-fetch

---

*[components/net/file_loader.rs, line 111 \[r9\]](https://reviewable.io:443/reviews/servo/servo/10961#-KHz5V0sAjNqi9m82ajO:-KHz5V0sAjNqi9m82ajP:1325985269) ([raw file](https://github.com/servo/servo/blob/dd4baae890fa6a22c1e33f779571f02f6c77af9e/components/net/file_loader.rs#L111)):*
> ```Rust
>         };
> 
>         let (msg_sender, msg_receiver) = ipc::channel().unwrap();
> ```

We don't need this, since it's not an HTTP request.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10961#-:-KHz6-5qbiyl79X6rlOw:680642058)*

**Comment by creativcoder** [comment]
@jdm  I added the the receiving code, in receive_event() and `handle event()`. and i debugged and found; its getting stuck at `ImportScripts` method only, but the initial `new Worker()` invocation of `load_whole_resource()` in `run_workers_scope` works fine. 

Under `ImportScripts`, i found that the code never receives a network event, even when its added in `receive_event()` matching, and due to this it can't send `None` to the network side, which results it not being able to complete the ImportScript() within worker, which results in failing of tests. What am i missing in `workerglobalscope` event handling ?

**Comment by jdm** [comment]
Oh right. Ugh. Just make ImportScripts provide a load origin that doesn't have a request source; it's a synchronous method, so the event loop that would receive the request for a custom response doesn't get a change to run until we return from load_whole_resource, so we deadlock.

**Comment by jdm** [comment]
The reason the the initiate load_whole_resource for `new Worker()` works is that it runs in a new thread, so the event processing code gets to run on the main script thread.

**Comment by creativcoder** [comment]
@jdm  why do most of the tests also timeout even on `master` branch of servo on my local machine, which do not have my changes. That should run fine, but even that timeouts.

**Comment by jdm** [comment]
I have not run into that before. What do you mean by "most of the tests", exactly?

**Comment by jdm** [comment]
That's a known intermittent failure covered by #11100.

**Comment by creativcoder** [comment]
Ah, I see. But buildbot logs now shows that, all tests are passing. So how do we make that pass ?  As for talking about, "many tests failing", it starts with:

```
not ready to composite: NotReadyToPaintImage(LayerHasOutstandingPaintMessages
 composition_request is already CompositeNow(NewFrameTree)
  │ composition_request is already CompositeNow(NewFrameTree)
  │ not ready to composite: NotReadyToPaintImage(JustNotifiedConstellation)
  │ got ready to save image query, result is DocumentLoading
  │ resetting ready_to_save_state!
  │ composition_request is already CompositeNow(NewFrameTree)
  │ sent response
  │ not ready to composite: NotReadyToPaintImage(JustNotifiedConstellation)
  │ got ready to save image query, result is Ready
  │ sent response
  │ ready to save image!
  │ composition_request is already CompositeNow(NewFrameTree)
  │ was ready to save, resetting ready_to_save_state
  └ Shutting down the Constellation after generating an output file or exit flag specified

 ▶ TIMEOUT [expected PASS] /html/dom/elements/global-attributes/dir_auto-contained-textarea-R.html
 │ 
 └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/img_simple.html

  ▶ TIMEOUT [expected PASS] /_mozilla/css/img_width_style_intrinsic_width_a.html

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_height_a.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_insertion_point_empty_a.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_line_height.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_selection_a.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_placeholder.html
  │ 
  │ Shutting down the Constellation after generating an output file or exit flag specified
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_selection_ref.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/input_whitespace.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  ▶ TIMEOUT [expected PASS] /_mozilla/css/legacy_input_size_attribute_override_a.html
  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified

  │ 
  └ Shutting down the Constellation after generating an output file or exit flag specified
```

and continues, for subsequent tests. I have tried running it both on `master` as well as this branch, but the same timeouts occur, on many iterations of tests on my machine.

**Comment by nox** [comment]
@bors-servo retry

Witness me Homu.

**Comment by nox** [comment]
Closing this to see if it makes things better.

**Comment by nox** [comment]
@bors-servo clean r- try-

Which keywords will make you happy Homu?

**Comment by nox** [comment]
Closing just so that we clear things up.

**Comment by KiChjang** [comment]
The answer is that the author needs to force push a new commit hash into his branch, this will refresh the homu state.

**Comment by nox** [comment]
@jdm Coincidence, or not an intermittent on this PR?

**Comment by jdm** [comment]
I'm assuming a string of coincidences.

**Comment by creativcoder** [comment]
@jdm @nox can we have a retry on this ? i rebased it again.

**Comment by creativcoder** [comment]
@jdm can we get this merged now ?

**Pull request merged**